pkg/cache/cache.go

```diff
@@ -0,0 +1,251 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Provides cache mechanism based on etags to lazily build, and/or cache
+// results from expensive operation such that those operations are not
+// repeated unnecessarily. The operations can be created as a tree, and
+// replaced dynamically as needed.
+//
+// # Dependencies and types of caches
+//
+// This package uses a source/transform/sink model of caches to build
+// the dependency tree, and can be used as follows:
+//   - [NewSource]: A source cache that recomputes the content every time.
+//   - [NewStaticSource]: A source cache that always produces the
+//     same content, it is only called once.
+//   - [NewTransformer]: A cache that transforms data from one format to
+//     another. It's only refreshed when the source changes.
+//   - [NewMerger]: A cache that aggregates multiple caches into one.
+//     It's only refreshed when the source changes.
+//   - [Replaceable]: A cache adapter that can ben atomically
+//     replaced with a new one, and saves the previous results in case an
+//     error pops-up.
+//
+// # Atomicity
+//
+// Most of the operations are not atomic/thread-safe, except for
+// [Replaceable.Replace] which can be performed while the objects
+// are being read.
+package cache
+
+import (
+	"fmt"
+	"sync/atomic"
+)
+
+// Result is the data returned from a call to a cache. It can either be
+// created with [NewResultData] if the call was a success, or
+// [NewResultError] if the call resulted in an error.
+type Result[T any] struct {
+	data T
+	etag string
```

Comment:
Can you go into detail about the format and how it should be generated/used? is it optional? I noticed in `handler.go` this etag value is ignored in some cases. Is the requirement for this etag the same as the http etag or are the requirements relaxed?
afaik with the json and proto marshalling cache, this is defined as the sha512 of the bytes. How about with the structs in the aggregator?

Edit: Okay I see that time is used as the etag in some places. This is probably not meant to be shared between threads (as mentioned in the header above). I wonder if we should use a different name to avoid potential confusion with the actual HTTP etag concept?

Reply:
To be fair, this is meant as a "cache identifier", which in many occurences can be either the etag of the downloaded document, or the etag of the document that we'll serve, so it happens to often be an actual http etag. It doesn't have to, but it's a string that should correspond to the version of an object. You probably noticed that I'm sometimes generating the etag using uuid as a way to say "it's a random piece of data, I can't version it". I don't know if the name is confusing or helpful :-/ I found it useful because we know what this represents, and apart from not having to be quoted like regular http tags, they have about the same set of constraints.

Reply:
I'd call it a CacheEntry, maybe?